arenarobot/service/processor/apriltag_detector.py

```diff
@@ -0,0 +1,178 @@
+        self.dist_params       = np.array(dist_params) # distortion parameters
+        self.at_detector       = None                  # apriltag detector object
+        self.flip_matrix       = None                  # flip matrix
+        self.apriltag_family   = apriltag_family       # tag family
+        self.apriltags         = apriltag_locations    # known apriltag poses (world coordinates in meters)
+
+        processor_type = (ArenaRobotServiceProcessorApriltagDetector
+                          .DEVICE_INSTANCE_PROCESSOR_TYPE)
+
+        super().__init__(
+            device_instance_processor_type=processor_type,
+            device_instance_prefix="apriltag_detector_",
+            **kwargs
+        )
+    '''
+    Set up Apriltag Detector object and define a bunch of stuff
+    '''
+    def setup(self):
+        # Initialize camera using /dev/video0 (works for USB cameras 
+        # but not Arducam-type cameras because the Raspberry Pi
+        # Foundation is sooooo silly)
+        self.cap = cv2.VideoCapture(0)
+
+        # subsample to increasing processing speed
+        # note: selected resolution affects camera parameters
+        self.cap.set(3, self.camera_resolution[0])
+        self.cap.set(4, self.camera_resolution[1])
+
+        # camera matrix
+        self.mtx = np.array([[self.params[0], 0.000000000000, self.params[2]],
+                             [0.000000000000, self.params[1], self.params[3]],
+                             [0.000000000000, 0.000000000000, 1.000000000000]])
+
+        # apriltag detector
+        self.at_detector = Detector(searchpath=['apriltags'],
+                                    families=self.apriltag_family,
+                                    nthreads=4,
+                                    quad_decimate=2.0,
+                                    quad_sigma=0.0,
+                                    refine_edges=1,
+                                    decode_sharpening=0.25,
+                                    debug=0)
+
+        self.flip_matrix = np.array([[1,  0,  0, 0],
+                                     [0, -1,  0, 0],
+                                     [0,  0, -1, 0],
+                                     [0,  0,  0, 1]], dtype=float)
+
+        super().setup()
+
+    '''
+    Capture a camera frame, detect apriltags, calculate pose, and
+    publish pose information to MQTT topic: realm/d/<user>/<device>/processors/pose_transformed
+    TODO: Implement Kalman Filter
+    TODO: figure out how this works with different users/devices (not just jpedraza/john-pi)
+    TODO: handle getting wrong representation (apriltag solver has two possible solutions)
+    TODO: handle multiple apriltags in frame at same time
+    '''
+    def fetch(self):
+        # get grayscale frame for apriltag detection
+        ret, frame = self.cap.read()
+        im_gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
```

Comment:
Is `BGR` as input specific to this camera?

Reply:
I think frames captured with OpenCV use BGR by default, so this should work for any camera. If I had a different microservice that was capturing frames, converting to grayscale, and then sending those frames to this microservice, I would probably have to add a parameter. All the processing from capture is happening in one service so keeping it like this is probably fine, right?

Reply:
OK, yeah keep it as is for now.